lib/ex_unit/lib/ex_unit/formatter.ex

```diff
@@ -200,25 +202,138 @@ defmodule ExUnit.Formatter do
     formatter.(:error_info, value)
   end
 
-  defp code_multiline(expr, padding) when is_binary(expr) do
-    String.replace(expr, "\n", "\n" <> String.duplicate(" ", padding))
+  defp code_multiline(expr, padding_size) when is_binary(expr) do
+    padding = String.duplicate(" ", padding_size)
+    String.replace(expr, "\n", "\n" <> padding)
   end
 
-  defp code_multiline(expr, padding) do
-    code_multiline(expr |> Macro.to_string, padding)
+  defp code_multiline(expr, padding_size) do
+    code_multiline(Macro.to_string(expr), padding_size)
   end
 
-  defp inspect_multiline(expr, padding, width) do
-    width = if width == :infinity, do: width, else: width - padding
-    expr
-    |> inspect(pretty: true, width: width)
-    |> String.replace("\n", "\n" <> String.duplicate(" ", padding))
+  defp inspect_multiline(expr, padding_size, width) do
+    padding = String.duplicate(" ", padding_size)
+    width = if width == :infinity, do: width, else: width - padding_size
+    inspect(expr, [pretty: true, width: width])
+    |> String.replace("\n", "\n" <> padding)
   end
 
   defp make_into_lines(reasons, padding) do
     padding <> Enum.join(reasons, "\n" <> padding) <> "\n"
   end
 
+  defp format_diff(struct, formatter) do
+    if_value(struct.left, fn left ->
+      if_value(struct.right, fn right ->
+        format_diff(left, right, formatter) || ExUnit.AssertionError.no_value
+      end)
+    end)
+  end
+
+  @doc """
+  Formats the difference between `left` and `right`.
+
+  Returns `nil` if they are not the same data type.
+  """
+  def format_diff(left, right, formatter_fun)
+
+  def format_diff(<<left::bytes>>, <<right::bytes>>, formatter) do
+    String.myers_difference(left, right)
+    |> Enum.map_join(&format_diff_fragment(&1, formatter))
+    |> String.replace("\n", "\n" <> @inspect_padding)
+  end
+
+  def format_diff(%name{} = left, %name{} = right, formatter) do
+    left = Map.from_struct(left)
+    right = Map.from_struct(right)
+    format_map_diff(left, right, inspect(name), formatter)
+  end
+
+  def format_diff(%_{}, %_{}, _formatter), do: nil
+
+  def format_diff(%{} = left, %{} = right, formatter) do
+    format_map_diff(left, right, "", formatter)
+  end
+
+  def format_diff(left, right, formatter)
+  when is_number(left) and is_number(right)
+  when is_tuple(left) and is_tuple(right)
+  when is_list(left) and is_list(right) do
+    format_diff(inspect(left), inspect(right), formatter)
+  end
+
+  def format_diff(_left, _right, _formatter), do: nil
+
+  defp format_map_diff(left, right, name, formatter) do
+    {surplus, altered} =
```

Comment:
Put the calculation of surplus, altered and missing into a separate function called `map_diff`. :)